File: server/src/schemas/orderStatus.ts
import { z } from "zod";

/**
 * Канонический enum статусов заказа — единый источник истины.
 * СИНХРОНИЗИРОВАН с docs/order-status.md и android OrderStatus.kt (ТЗ §FR-A04).
 * Любое изменение — сначала в docs/order-status.md, затем здесь и в Kotlin.
 */
export const ORDER_STATUS = [
  // happy path
  "NEW",
  "CONFIRMED",
  "PROCESSING",
  "SHIPPED",
  "DELIVERED",
  "COMPLETED",
  // терминальные
  "CANCELLED",
  "RETURN_REQUESTED",
  "RETURNED",
  "REFUNDED",
] as const;

export const orderStatusSchema = z.enum(ORDER_STATUS);
export type OrderStatus = z.infer<typeof orderStatusSchema>;

/** Статусы, в которых покупатель может отменить заказ (ТЗ §FR-B06). */
export const CANCELLABLE_BY_BUYER: readonly OrderStatus[] = ["NEW", "CONFIRMED", "PROCESSING"];

export function isCancellableByBuyer(status: OrderStatus): boolean {
  return CANCELLABLE_BY_BUYER.includes(status);
}

File: android/app/src/main/kotlin/com/wasat/shop/domain/model/OrderStatus.kt
package com.wasat.shop.domain.model

/**
 * Канонический enum статусов заказа — единый источник истины.
 * СИНХРОНИЗИРОВАН с docs/order-status.md и server schemas/orderStatus.ts (ТЗ §FR-A04).
 * Любое изменение — сначала в docs/order-status.md, затем здесь и в Zod-схеме.
 */
enum class OrderStatus {
    // happy path
    NEW,
    CONFIRMED,
    PROCESSING,
    SHIPPED,
    DELIVERED,
    COMPLETED,

    // терминальные
    CANCELLED,
    RETURN_REQUESTED,
    RETURNED,
    REFUNDED;

    /** Может ли покупатель отменить заказ в этом статусе (ТЗ §FR-B06). */
    val isCancellableByBuyer: Boolean
        get() = this == NEW || this == CONFIRMED || this == PROCESSING
}
